client/api.c

```diff
@@ -1332,7 +1332,204 @@ TDNFRepoSync(
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    /* create analog to SOLVABLE_REQUIRES */
+    idDepends = pool_str2id(pTdnf->pSack->pPool, TDNF_ID_DEPENDS, 1);
+    idRequiresPre = pool_str2id(pTdnf->pSack->pPool, TDNF_ID_REQUIRES_PRE, 1);
+
+    dwError = TDNFRefresh(pTdnf);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    dwError = SolvCreateQuery(pTdnf->pSack, &pQuery);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (!pRepoqueryArgs->nExtras)
+    {
+        if (!pRepoqueryArgs->nInstalled || pRepoqueryArgs->nAvailable)
+        {
+            dwError = SolvAddAvailableRepoFilter(pQuery);
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+
+        if (pRepoqueryArgs->nInstalled || pRepoqueryArgs->nDuplicates)
+        {
+            dwError = SolvAddSystemRepoFilter(pQuery);
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+    }
+
+    dwError = TDNFApplyScopeFilter(pQuery,
+        pRepoqueryArgs->nUpgrades ? SCOPE_UPGRADES : SCOPE_ALL);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (pRepoqueryArgs->pszSpec)
+    {
+        dwError = SolvApplySinglePackageFilter(pQuery, pRepoqueryArgs->pszSpec);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = SolvApplyListQuery(pQuery);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (pRepoqueryArgs->nExtras)
+    {
+        dwError = SolvApplyExtrasFilter(pQuery);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    else if (pRepoqueryArgs->nDuplicates)
+    {
+        dwError = SolvApplyDuplicatesFilter(pQuery);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (pRepoqueryArgs->ppszWhatDepends != NULL)
+    {
+        dwError = SolvApplyDepsFilter(pQuery, pRepoqueryArgs->ppszWhatDepends, idDepends);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    for (i = 0; i < REPOQUERY_KEY_COUNT; i++)
+    {
+        if (pRepoqueryArgs->pppszWhatKeys[i])
+        {
+            dwError = SolvApplyDepsFilter(pQuery,
+                        pRepoqueryArgs->pppszWhatKeys[i],
+                        allDepKeyIds[i]);
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+    }
+
+    if (pRepoqueryArgs->pszFile)
+    {
+        dwError = SolvApplyFileProvidesFilter(pQuery, pRepoqueryArgs->pszFile);
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    dwError = SolvGetQueryResult(pQuery, &pPkgList);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (pRepoqueryArgs->nChangeLogs)
+    {
+        dwError = TDNFPopulatePkgInfoArray(pTdnf->pSack, pPkgList, DETAIL_CHANGELOG, &pPkgInfo, &dwCount);
+    }
+    else if (pRepoqueryArgs->nSource)
+    {
+        dwError = TDNFPopulatePkgInfoArray(pTdnf->pSack, pPkgList, DETAIL_SOURCEPKG, &pPkgInfo, &dwCount);
+    }
+    else
+    {
+        dwError = TDNFPopulatePkgInfoArray(pTdnf->pSack, pPkgList, DETAIL_LIST, &pPkgInfo, &dwCount);
+    }
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    if (pRepoqueryArgs->nDepends)
+    {
+        dwError = TDNFPopulatePkgInfoArrayDependencies(
```

Comment:
I found no easy way, but while pondering about it I found that the logic can be made much simpler because there is only one dependency key allowed. It's much cleaner now.